spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/src/test/java/com/example/ApplicationTests.java

```diff
@@ -0,0 +1,149 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.example;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.cloud.logging.LogEntry;
+import com.google.cloud.logging.Logging;
+import com.google.cloud.logging.LoggingOptions;
+import com.google.cloud.logging.Payload.StringPayload;
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.cloudtrace.v1.GetTraceRequest;
+import com.google.devtools.cloudtrace.v1.Trace;
+import com.google.devtools.cloudtrace.v1.TraceServiceGrpc;
+import com.google.devtools.cloudtrace.v1.TraceServiceGrpc.TraceServiceBlockingStub;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.auth.MoreCallCredentials;
+import org.awaitility.Duration;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.boot.web.server.LocalServerPort;
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assume.assumeThat;
+
+/**
+ * Verifies that the logged Traces on the sample application appear in StackDriver.
+ *
+ * @author Daniel Zou
+ */
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT, classes = { Application.class })
+public class ApplicationTests {
+
+	@LocalServerPort
+	private int port;
+
+	@Autowired
+	private GcpProjectIdProvider projectIdProvider;
+
+	@Autowired
+	private CredentialsProvider credentialsProvider;
+
+	private String url;
+
+	private TestRestTemplate testRestTemplate;
+
+	private Logging logClient;
+
+	private TraceServiceBlockingStub traceServiceStub;
+
+	@BeforeClass
+	public static void checkToRun() {
+		assumeThat(
+				"Google Cloud Trace integration tests are disabled. "
+						+ "Please use '-Dit.trace=true' to enable them. ",
+				System.getProperty("it.trace"), is("true"));
+	}
+
+	@Before
+	public void setupTraceClient() throws IOException {
+		this.url = String.format("http://localhost:%d/", this.port);
+		this.testRestTemplate = new TestRestTemplate();
```

Comment:
One more time: you don't need to instantiate this object and do some hard formatting logic for the port.
There is just enough to `@Autowired` this object from the application context.
See `ApplicationTests` in the `spring-cloud-gcp-data-datastore-sample`.

Reply:
That's how we had it before, but that `TestRestTemplate` was getting instrumented by Sleuth and interfered with the test.
@dzou, maybe add a comment above this line explaining this.

Reply:
No, this is a special case with the Trace sample test. Due to the configuration of the sample app, the autowired instance of TestRestTemplate has its own built-in Trace instrumentation and will attach a random `xb-3-traceid` header automatically to all requests. The traces on Stackdriver will use this ID. However, in the test we want to set the traceid manually with `x-cloud-trace-context` so we can use the value to look it up and verify it on StackDriver.

So here I have to create a new rest template instance which does not have the configuration that automatically adds that xb-3 header.

Reply:
Uh! So, sounds like you want to overcome (mock ?) the real functionality with your own code.
Why just not test whatever is provided by the tracer automatically?

How it can be an integration test if we don't use real functionality?

Maybe we can just add some interceptor into the autowired `TestRestTemplate` to catch that generated header and then verify against it in the end of test?

Thanks for understanding

Reply:
Hmm, so the only functionality we are changing is overriding the traceId set in the request with a non-random one. That lets us know the traceId value beforehand so we can check it was successfully registered in StackDriver.

So I think it is ok here for the integration test since we are not modifying the core logic of what is going on. If there were some way of determining the random ID generated for the requests by the rest template beforehand, we could eliminate doing this. This was the only way that Mike and I could figure out how to make the traceId known by the test.

Reply:
OK. Fair enough. No reason to argue anymore.

Please, add an in-line comment here to explain "Why?" and it is going to be good for merging!

Reply:
Sounds good, thanks.